l3kernel/l3bitset.dtx

```diff
@@ -0,0 +1,551 @@
+      {
+        \int_compare:nNnTF {\str_count:N #1 } < { #2 + 1 }
+          {
+            %extend the str
+            \int_step_inline:nn { #2 - \str_count:N #1 }
+              {
+                \str_gput_left:Nn #1 { 0 }
+              }
+            \str_gput_left:Nn #1 { 1 }
+          }
+          {
+            % replace the value
+            \str_gset:Nx #1
+              {
+                \str_range:Nnn #1 {1}{-2 -(#2)}
+                1
+                \str_range:Nnn #1 {0 -(#2)}{-1}
+              }
+          }
+      }
+  }
+\cs_generate_variant:Nn \bitset_set:Nn  {cn}
+\cs_generate_variant:Nn \bitset_gset:Nn {cn}
+%    \end{macrocode}
+% \end{macro}
+
+% \begin{macro}
+%   {
+%     \bitset_clear:Nn,  \bitset_clear:cn,
+%     \bitset_gclear:Nn, \bitset_gclear:cn
+%   }
+% Clearing a bit has only to do something if the string is longer than then index.
+%    \begin{macrocode}
+\cs_new_protected:Npn \bitset_clear:Nn #1 #2  %#1 name, #2 index (zero based)
+ {
+   \int_compare:nNnT {#2 } > { -1 }
+     {
+       \int_compare:nNnT {\str_count:N #1 } > { #2 }
+         {
+           % need to replace the str
+           \str_set:Nx #1
+             {
+               \str_range:Nnn #1 {1}{-2 - (#2)}
+                0
+               \str_range:Nnn #1 { 0 - (#2)}{-1}
+             }
+         }
+     }
+ }
+
+\cs_new_protected:Npn \bitset_gclear:Nn #1 #2  %#1 name, #2 index (zero based)
+ {
+   \int_compare:nNnT {#2 } > { -1 }
+     {
+       \int_compare:nNnT {\str_count:N #1 } > { #2 }
+         {
+           % need to replace the str
+           \str_gset:Nx #1
+             {
+               \str_range:Nnn #1 {1}{-2 -(#2)}
+                0
+               \str_range:Nnn #1 {0 - (#2)}{-1}
+             }
+         }
+     }
+ }
+\cs_generate_variant:Nn \bitset_clear:Nn {cn}
+\cs_generate_variant:Nn \bitset_gclear:Nn {cn}
+%    \end{macrocode}
+%  \end{macro}
+% \begin{macro}
+%   {
+%     \bitset_to_arabic:N, \bitset_to_arabic:c,
+%     \bitset_to_bin:N,    \bitset_to_bin:c,
+%   }
+%   The naming of the commands follow the names in the \texttt{int} module.
+%   \cs{bitset_to_arabic:N} uses \cs{int_from_bin:n} and so can only handle small
+%   numbers. There is no test for the string length, so that long bitset with
+%   many leading zeros work nevertheless.
+%    \begin{macrocode}
+\cs_new:Npn \bitset_to_arabic:N #1
+  {
+    \exp_args:Ne \int_from_bin:n { #1 }
+  }
+
+\cs_new:Npn \bitset_to_bin:N #1
+  {
+    #1
+  }
+
+\cs_generate_variant:Nn \bitset_to_arabic:N  {c}
+\cs_generate_variant:Nn \bitset_to_bin:N {c}
+%    \end{macrocode}
+% \end{macro}
+% \begin{macro}
+%   {
+%     \bitset_index:Nn, \bitset_index:cn
+%   }
+%    \begin{macrocode}
+\cs_new:Npn \bitset_index:Nn #1 #2
+ {
+   \int_eval:n { \str_item:Nn #1 { -1 - (#2)} +0 }
+ }
+\cs_generate_variant:Nn \bitset_index:Nn {cn}
+%    \end{macrocode}
+% \end{macro}
+
+% \begin{macro}
+%   {
+%     \bitset_show:N, \bitset_show:c,
+%     \bitset_log:N,  \bitset_log:c
```

Comment:
As for `\bitset_to_bin`, I would like a fixed length (or maybe minimum length) `:Nn` version for these.

Reply:
I have no real opinion what a user expect or need for binary representation, (and if someone actually need it) I mostly used it to check if the index were correctly set. Do you know real use cases here?

Reply:
I think we agreed at the team meeting that this is likely best left 'open ended'.